### ISSUE TITLE
[4.0] very slow upload of video frames in GPU

### ISSUE THREAD

**Issue by frankiezafe** [godotengine/godot#57147]
### Godot version

6a3ff8fa1f2cbad5752cf039827206ac265f68f6

### System information

Linux Mint 20.1, nvidia-driver-470

### Issue description

Hello,

I'm currently working on the ffmpeg module for godot 4.0 (see https://gitlab.com/polymorphcool/godot-ffmpeg/-/tree/4.0)

There's a massive performance drop when i try to upload HD textures. Textures < 1024x1024 are pushed seamlessly. As soon I try 1920x1080, frame rate drops around 1 fps, majority of the time is lost during the call to `ImageTexture::update(const Ref<Image> &p_image)`.

see https://gitlab.com/polymorphcool/godot-ffmpeg/-/blob/4.0/fast_forward_playback.cpp#L839

The code is inspired by the theora module.

### Steps to reproduce

- clone git@gitlab.com:polymorphcool/godot-ffmpeg.git into modules/ffmpeg/
- check out the 4.0 branch
- compile
- create a empty project
- add a videoplayer and put an HD mp4 stream in it
- play

**Comment by Calinou** [comment]
Can you reproduce this on 3.4.2?

**Comment by frankiezafe** [comment]
On the 3.4.2, i was using `texture->set_data(img);` see: https://gitlab.com/polymorphcool/godot-ffmpeg/-/blob/master/fast_forward_playback.cpp#L757
Performances were better, meaning i could decompress & upload 3 1920x1080 frames in parallel without significant drop. But as soon i was going to 2k, i had the same effect.

Is it due to an automatism in the code: above a certain resolution, mipmaps are enabled by default for instance?

For the ProxyTexture-style, I'm ok to work on it, i just need some guidelines. I guess i'll have to extends the drivers. This could be super usefull to upload data in particle system and complex shaders via texture.

**Comment by Calinou** [comment]
> Is it due to an automatism in the code: above a certain resolution, mipmaps are enabled by default for instance?

No, Godot does not automatically enable mipmap generation in the Image class.

**Comment by underdoeg** [comment]
I am seeing the same thing with a custom video player module built around mpv. Everything has an alright performance but texture->update is a huge bottleneck.
